llvm: Support work for using versions 9 and 10-dev

vtjnash opened the pull request.

vtjnash committed:
llvm: Support work for using versions 9 and 10-dev

vchuravy committed:
update to LLVM10 changes

vchuravy reviewed (approved):
Checked on Linux for ToT and 9.0.0

vchuravy commented on src/llvm-api.cpp:
```diff
@@ -176,37 +176,6 @@ LLVMExtraCreateFunctionPass(const char *Name, jl_value_t *Callback)
-};
-
-extern "C" JL_DLLEXPORT LLVMPassRef
-LLVMExtraCreateBasicBlockPass(const char *Name, jl_value_t *Callback)
```
CC: @maleadt

maleadt replied:
I guess it's a bit sudden to get rid of it like that, but apparently the BB pass is broken and I don't use it, so :man_shrugging:

vtjnash replied:
Maybe sudden, but been deprecated for a while though.

vchuravy commented on src/jitlayers.cpp:
```diff
@@ -305,28 +312,8 @@ void JuliaOJIT::DebugObjectRegistrar::registerObject(RTDyldObjHandleT H, const O
     const ObjT& Object = Obj;
 #endif
 
-    object::OwningBinary<object::ObjectFile> SavedObject = LO->getObjectForDebug(*Object);
```
Do we no longer need to do this? Or is it incompatible with LLVM10?

vtjnash replied:
LLVM doesn't want us to do this, and has been making it increasingly difficult to avoid making these several copies. We used to have private copies of the llvm::DebugObjectRegistrar code to avoid these copies, but have since removed it (so this code served no real purpose). Unless we planning to convince LLVM to use `std::unique_ptr` less, the current strategy for C++ manual memory management now seems to be just to make private copies of everything. The things we do to avoid having a garbage collector...

vtjnash commented:
Note: need to remember to change LICENSE.md too

vtjnash commented:
@staticfloat Can you make an Yggdrasil run with LLVM 9.0.1—once that is released—and we can upgrade it here just after we branch for v1.4?

vtjnash committed:
internals: redesign debug object slide handling

LLVM redesigned their debug object slide handling (again), and added
asserts for the old, less good way of doing things. Thus we needed to
make the same changes (and which makes us more correct, like LLVM).
